### PR TITLE
Improve HOWTOs

### PR DESCRIPTION
**Pull request by avital**
1. Render HOWTOs as nicely formatted inline diffs in docs.
   These diffs are syntax highlighted for both Python and diffs

2. Simplify the multi-device parallel training HOWTO

3. Re-pack the ensembling diff. (Once I applied it and re-packed it
   it got a lot shorter… I wonder if this is something we need to do
   for every commit?)

4. Make the HOWTO docs page directly show HOWTOs, and separate the
   mechanics of the HOWTO system into a separate “HOWTOs HOWTO” page.

5. Modify the pack_howto_diffs.sh script to only pack changes in
   examples/. This significantly simplified the process of iterating
   on HOWTO diffs.

### CODE REVIEW

File: docs/Makefile [hunk 1]
```diff
@@ -17,4 +17,5 @@ help:
 # Catch-all target: route all unknown targets to Sphinx using the new
 # "make mode" option.  $(O) is meant as a shortcut for $(SPHINXOPTS).
 %: Makefile
+	python3 format_howtos.py
```
Comment (jheek) [inline docs/Makefile]: I don't think travis uses the make file currently it probably should know that it actually does something more than calling sphinx

File: docs/format_howtos.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
+"""
```
Comment (jheek) [inline docs/format_howtos.py]: How difficult is it to incorporate this into sphinx? Ideally we would load this code as a sphinx plugin and use something like `howto:: [name]` in the rst files.

File: docs/format_howtos.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    # Run `diff` through the normal pygments Python syntax
+    # highlighter.  Get back an array of HTML lines
+    colored_diff = (
+        pygments.highlight('\n'.join(diff), PythonLexer(), RawHtmlFormatter())
+    ).splitlines()
+
+    # Write a newly formatted diff-and-Python syntax highlighted
+    # output HTML file, that can be inline included into any .rst
+    # file.
+    with open(output_file, "w") as out_file:
```
Comment (jheek) [inline docs/format_howtos.py]: nit: " -> '

File: docs/format_howtos.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+"""
+
+import pygments
+import pygments.formatters
+from pygments.lexers import PythonLexer
+
+import os
+
+def main():
+    for diff_filename in os.listdir(os.path.join('..', 'howtos', 'diffs')):
```
Comment (jheek) [inline docs/format_howtos.py]: nit: this file has 4 spaces instead of 2

File: docs/format_howtos.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        format_howto(os.path.join('..', 'howtos', 'diffs', diff_filename),
+                     os.path.join('_formatted_howtos', diff_filename + '.html'))
+
+def format_howto(input_file, output_file):
+    # Load one of our HOWTO diff files.
+    with open(input_file) as f:
+        diff = f.readlines()
```
Comment (jheek) [inline docs/format_howtos.py]: I guess we load the raw .diff file here directly? Would git produce different diffs due to its merging strategy?
Especially the line numbers will quickly bit rot when other parts of the code are updated without causing conflicts.

File: docs/Makefile (hunk 1 above)
Comment (avital) [reply docs/Makefile]: OK I'll look at the travis script as well.

File: docs/format_howtos.py (hunk 2 above)
Comment (avital) [reply docs/format_howtos.py]: I was thinking about that as well. Honestly I don't know, but the approach in this CL has the added benefit that it's easy to debug/test the format_howto script. If it were entirely in Sphinx I could see that being harder. This version is more like "unix standalong executables that combine via pipes".

So, given that this approach works, doesn't have obvious flaws and allows for testing/debugging, I'd like to err on the side of "let's keep it the way it is".

File: docs/format_howtos.py (hunk 3 above)
Comment (avital) [reply docs/format_howtos.py]: Acknowledged.

File: docs/format_howtos.py (hunk 4 above)
Comment (avital) [reply docs/format_howtos.py]: Acknowledged.

File: docs/format_howtos.py (hunk 5 above)
Comment (avital) [reply docs/format_howtos.py]: Yes the line number will rot, but git diffs keep metadata about which commit they were forked from which (barring history rewriting on prerelease which we shouldn't do) allows them to be applied later as well. (This is how Git 3 way merge works)

But we don't show the line numbers directly to anyone.

What is true is that at times when the base examples change, then the diff here will be longer than it should be (imagine we brought in part of a HOWTO into the base example). At that point you would want to re-apply and re-pack all of the HOWTOs to make the cleaner.

We could consider doing that as a hook.

**Comment by avital** [comment]
I pushed a new commit (hopefully) resolving your concerns, please take another look

File: docs/format_howtos.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    diff = f.readlines()
+
+  # Ignore the first two line, which look like:
+  #
+  # diff --git a/examples/mnist/train.py b/examples/mnist/train.py
+  # index 51d2fde..a9d7dcb 100644
+  diff = diff[2:]
+
+  # Remove more than one empty line in a row
+  diff = [diff[lineno] for lineno in range(len(diff))
```
Comment (marcvanzee) [inline docs/format_howtos.py]: Are you trying to remove double newlines? It is not entirely clear to me from the comment. Perhaps it is easier to understand if you use a regex? https://stackoverflow.com/a/28902081

File: docs/format_howtos.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        diff[lineno-1].rstrip(' \n') == ''
+       )]
+
+  # Don't do any special formatting
+  class RawHtmlFormatter(pygments.formatters.HtmlFormatter):
+    def wrap(self, source, outfile):
+      return source
+
+  # Run `diff` through the normal pygments Python syntax
+  # highlighter.  Get back an array of HTML lines
```
Comment (marcvanzee) [inline docs/format_howtos.py]:
```suggestion
  # highlighter. Get back an array of HTML lines.
```

File: docs/format_howtos.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  diff = diff[2:]
+
+  # Remove more than one empty line in a row
+  diff = [diff[lineno] for lineno in range(len(diff))
+      if lineno == 0 or not (
+        diff[lineno].rstrip(' \n') == '' and
+        diff[lineno-1].rstrip(' \n') == ''
+       )]
+
+  # Don't do any special formatting
```
Comment (marcvanzee) [inline docs/format_howtos.py]:
```suggestion
  # Don't do any special formatting.
```

File: docs/format_howtos.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  # highlighter.  Get back an array of HTML lines
+  colored_diff = (
+    pygments.highlight('\n'.join(diff), PythonLexer(), RawHtmlFormatter())
+  ).splitlines()
+
+  # Write a newly formatted diff-and-Python syntax highlighted
+  # output HTML file, that can be inline included into any .rst
+  # file.
+  with open(output_file, 'w') as out_file:
+    # Add the relevant DIVs that Sphinx adds around code blocks
```
Comment (marcvanzee) [inline docs/format_howtos.py]:
```suggestion
    # Add the relevant DIVs that Sphinx adds around code blocks.
```

File: docs/howtos-howto.md [hunk 10]
```diff
@@ -64,14 +50,16 @@ The workflow script for the Github action can be found at
 
 ## Resolving a conflict with an existing HOWTO
 
-> :warning: TODO: This is still in progress. Currently conflicts are resolved by
-            checking whether the Github action fails, and if so, manually making
-            the required fixes.
+TODO: This is still in progress. Currently conflicts are resolved by
+checking whether the Github action fails, and if so, manually making
+the required fixes.
+</div>
```
Comment (marcvanzee) [inline docs/howtos-howto.md]: Remove this?

File: docs/howtos.rst [hunk 11]
```diff
@@ -0,0 +1,53 @@
+Flax HOWTOs
+===========
+
+Flax aims to be a thin library of composable primitives. You compose
+these primites yourself into a training loop that you write and is fully under
+your control. You have full freedom to modify the behavior of your training loop,
+and you should generally not need special library support to implement
+the modifications you want.
+
+To help you get started, we show some sample diffs, which
```
Comment (marcvanzee) [inline docs/howtos.rst]: This newline seems wrong, will it render correctly?

File: docs/howtos.rst [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+we call "HOWTOs". These HOWTOs show common modifications to training loops. For instance,
+the HOWTO for ensembling learning demonstrates what changes should be made to
+the standard MNIST example in order to train an ensemble of models on
+multiple devices.
+
+Note that these HOWTOs do not require special library support, they just
+demonstate how assembling the JAX and Flax primitives in different ways
+allow you to achieve various training loop modifications.
```
Comment (marcvanzee) [inline docs/howtos.rst]:
```suggestion
allows you to achieve various training loop modifications.
```

Also, "achieve various training loop modifications" sounds a bit strange. I don't think you "achieve" a modification, you "make" it.

File: docs/format_howtos.py (hunk 6 above)
Comment (avital) [reply docs/format_howtos.py]: Well the file is already split into lines, so you can't use a regexp. (And the lines aren't actually empty, they contain the first space that designates whether it's an addition or removal of a line, or just context).

I'll modify the comment.

File: docs/howtos-howto.md (hunk 10 above)
Comment (avital) [reply docs/howtos-howto.md]: Oh? What fixed this? Do we have this test set up on Travis?

File: docs/howtos.rst (hunk 11 above)
Comment (avital) [reply docs/howtos.rst]: I'm not sure what newline you mean. Is it that the line is shorter than 80c? Single newlines don't have any meaning in restructuredText. But two newlines in a row are a paragraph delineator?

File: docs/format_howtos.py (hunk 6 above)
Comment (avital) [reply docs/format_howtos.py]: I pushed another change rewriting the logic a bit and adding a clarifying comment. Please take another look.

File: docs/howtos.rst (hunk 12 above)
Comment (avital) [reply docs/howtos.rst]: You're right, I fixed that.

File: docs/format_howtos.py (hunk 6 above)
Comment (marcvanzee) [reply docs/format_howtos.py]: Thanks for the clarifying comment!

File: docs/howtos-howto.md (hunk 10 above)
Comment (marcvanzee) [reply docs/howtos-howto.md]: No I mean removing the "</div>", this is markdown, right?

File: docs/howtos.rst (hunk 11 above)
Comment (marcvanzee) [reply docs/howtos.rst]: Indeed I meant the newline on the line that is shorter than 80c. SG.

File: docs/howtos-howto.md (hunk 10 above)
Comment (avital) [reply docs/howtos-howto.md]: Oh oops, yes! Done.
